components/iam/pkg/oidc/router.go

```diff
@@ -5,16 +5,149 @@
 package oidc
 
 import (
-	"github.com/go-chi/chi/v5"
+	"context"
+	"encoding/json"
 	"net/http"
+	"time"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"golang.org/x/oauth2"
+
+	"github.com/go-chi/chi/v5"
 )
 
-func Router() *chi.Mux {
-	router := chi.NewMux()
+func Router(oidcService *OIDCService) *chi.Mux {
+	router := chi.NewRouter()
 
-	router.HandleFunc("/start", func(writer http.ResponseWriter, request *http.Request) {
-		writer.Write([]byte(`hello`))
+	router.Route("/start", func(r chi.Router) {
+		r.Use(oidcService.clientConfigMiddleware())
+		r.Get("/", oidcService.getStartHandler())
+	})
+	router.Route("/callback", func(r chi.Router) {
+		r.Use(oidcService.clientConfigMiddleware())
+		r.Use(OAuth2Middleware)
+		r.Get("/", oidcService.getCallbackHandler())
 	})
 
 	return router
 }
+
+type keyOIDCClientConfig struct{}
+
+const (
+	stateCookieName = "state"
+	nonceCookieName = "nonce"
+)
+
+func (oidcService *OIDCService) getStartHandler() http.HandlerFunc {
+	return func(rw http.ResponseWriter, r *http.Request) {
+		log.Trace("at start handler")
+
+		ctx := r.Context()
+		config, ok := ctx.Value(keyOIDCClientConfig{}).(OIDCClientConfig)
+		if !ok {
+			http.Error(rw, "config not found", http.StatusInternalServerError)
+			return
+		}
+
+		startParams, err := oidcService.GetStartParams(&config)
+		if err != nil {
+			http.Error(rw, "failed to start auth flow", http.StatusInternalServerError)
+			return
+		}
+
+		http.SetCookie(rw, newCallbackCookie(r, nonceCookieName, startParams.Nonce))
+		http.SetCookie(rw, newCallbackCookie(r, stateCookieName, startParams.State))
+
+		http.Redirect(rw, r, startParams.AuthCodeURL, http.StatusTemporaryRedirect)
+	}
+}
+
+func newCallbackCookie(r *http.Request, name string, value string) *http.Cookie {
+	return &http.Cookie{
+		Name:     name,
+		Value:    value,
+		MaxAge:   int(10 * time.Minute.Seconds()),
+		Secure:   r.TLS != nil,
+		SameSite: http.SameSiteLaxMode,
+		HttpOnly: true,
+	}
+}
+
+// The config middleware is responsible to retrieve the client config suitable for request
+func (oidcService *OIDCService) clientConfigMiddleware() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			log.Trace("at config middleware")
+
+			config, err := oidcService.GetClientConfigFromRequest(r)
+			if err != nil {
+				log.Warn("client config not found: " + err.Error())
+				http.Error(rw, "config not found", http.StatusNotFound)
+				return
+			}
+
+			ctx := context.WithValue(r.Context(), keyOIDCClientConfig{}, config)
+			next.ServeHTTP(rw, r.WithContext(ctx))
```

Comment:
I'd recommend having dedicated helpers for this:

```golang
func OIDCClientConfigToContext(ctx context.context, config) context.Context {.. }

func OIDCClientCOnfigFromContext(ctx context..) {..}
```

What this does is encapsulated access to the particualr context key in public methods. That means you make the way to access the key private, and as a result can change it.

Reply:
good point